### PR TITLE
Implement OnFailurePolicy in Propeller

### PR DESCRIPTION
**Pull request by EngHabu**
# TL;DR
Add support to OnFailurePolicy defined at the Workflow/Subworkflow levels. It defaults to `FAIL_IMMEDIATELY` keeping current behavior the same.

## Type
 - [ ] Bug Fix
 - [X] Feature
 - [ ] Plugin

## Are all requirements met?

 - [X] Code completed
 - [ ] Smoke tested
 - [ ] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
When set to `FAIL_AFTER_EXECUTABLE_NODES_COMPLETE`, it leads to the workflow continuing to run until all nodes have either succeeded, failed, timedout or skipped.

Screen Shot 2020-06-15 at 1 54 41 PM

## Tracking Issue
https://github.com/lyft/flyte/issues/191

### CODE REVIEW

**Comment by kumare3** [comment]
Or description update

File: pkg/controller/nodes/executor.go [hunk 1]
```diff
@@ -602,7 +613,10 @@ func canHandleNode(phase v1alpha1.NodePhase) bool {
 		phase == v1alpha1.NodePhaseSucceeding
 }
 
-func (c *nodeExecutor) RecursiveNodeHandler(ctx context.Context, execContext executors.ExecutionContext, dag executors.DAGStructure, nl executors.NodeLookup, currentNode v1alpha1.ExecutableNode) (executors.NodeStatus, error) {
+func (c *nodeExecutor) RecursiveNodeHandler(ctx context.Context, execContext executors.ExecutionContext,
```
Comment (kumare3) [inline pkg/controller/nodes/executor.go:616]: I think you will have to change CanExecute, add support for upstream failed state, then just fail the downstream node

File: pkg/controller/nodes/executor.go [hunk 2]
```diff
@@ -535,18 +537,24 @@ func (c *nodeExecutor) handleDownstream(ctx context.Context, execContext executo
...
 		}
-		if state.HasFailed() {
-			logger.Debugf(ctx, "Some downstream node has failed, %s", state.Err)
-			return state, nil
-		}
-		if state.HasTimedOut() {
-			logger.Debugf(ctx, "Some downstream node has timedout")
-			return state, nil
+
+		if state.HasFailed() || state.HasTimedOut() {
```
Comment (kumare3) [inline pkg/controller/nodes/executor.go:546]: also similar check in subworkflow?

**Comment by kumare3** [comment]
I am thinking its not ready yet right?

**Comment by EngHabu** [comment]
Ready for review @kumare3

File: pkg/apis/flyteworkflow/v1alpha1/workflow.go [hunk 3]
```diff
@@ -191,6 +191,10 @@ type WorkflowSpec struct {
 	// refer to specific outputs of a subset of the nodes executed in the Workflow. When executing the end-node,
 	// the execution engine will traverse these bindings and assemble the final set of outputs of the workflow.
 	OutputBindings []*Binding `json:"outputBindings,omitempty"`
+
+	// Defines the policy for handling failures whether it's to fail immediately, or let the nodes run
+	// to completion.
+	OnFailurePolicy core.WorkflowMetadata_OnFailurePolicy `json:"onFailurePolicy,omitempty"`
```
Comment (kumare3) [inline pkg/apis/flyteworkflow/v1alpha1/workflow.go:197]: should we wrap this in a struct?

Comment (kumare3) [reply pkg/apis/flyteworkflow/v1alpha1/workflow.go:197]: like
```
FailurePolicy{
enum
... future
}```

This also helps in jsonpb

File: pkg/apis/flyteworkflow/v1alpha1/workflow.go [hunk 4]
```diff
@@ -222,6 +226,10 @@ func (in *WorkflowSpec) FromNode(name NodeID) ([]NodeID, error) {
 	return downstreamNodes, nil
 }
 
+func (in *WorkflowSpec) GetOnFailurePolicy() core.WorkflowMetadata_OnFailurePolicy {
```
Comment (kumare3) [inline pkg/apis/flyteworkflow/v1alpha1/workflow.go:229]: we need to use jsonpb to deserialize.

File: pkg/controller/nodes/executor.go (hunk 2 above)
Comment (kumare3) [reply pkg/controller/nodes/executor.go:546]: aah you are right

File: pkg/apis/flyteworkflow/v1alpha1/workflow.go (hunk 3 above)
Comment (EngHabu) [reply pkg/apis/flyteworkflow/v1alpha1/workflow.go]: done

File: pkg/apis/flyteworkflow/v1alpha1/workflow.go (hunk 4 above)
Comment (EngHabu) [reply pkg/apis/flyteworkflow/v1alpha1/workflow.go]: Do we need jsonpb? this should work, right?

File: pkg/apis/flyteworkflow/v1alpha1/iface.go [hunk 5]
```diff
@@ -153,6 +153,25 @@ func (b BranchNodePhase) String() string {
 	return "Undefined"
 }
 
+//go:generate enumer --type WorkflowOnFailurePolicy --json --yaml --trimprefix=WorkflowOnFailurePolicy
+
+// Failure Handling Policy
+type WorkflowOnFailurePolicy int32
```
Comment (kumare3) [inline pkg/apis/flyteworkflow/v1alpha1/iface.go:159]: Interesting, so if we add one more element, we will need to add it here?

Comment (kumare3) [reply pkg/apis/flyteworkflow/v1alpha1/iface.go:159]: I think I like this, but it does not follow our other enums right?

**Comment by kumare3** [comment]
Looks good to me, I think we need to agree that all top level enums will be handled so.

File: pkg/apis/flyteworkflow/v1alpha1/iface.go (hunk 5 above)
Comment (EngHabu) [reply pkg/apis/flyteworkflow/v1alpha1/iface.go]: Is this better?
